[LogsUI] [InfraUI] Turn source configuration into a tab and standardize the main navigation

Kerry350 opened the pull request:
## Summary

This ticket implements #39071, and also implements the Infra focussed concerns of [#150](https://github.com/elastic/observability-dev/issues/150) by changing the main navigation to align with SIEM / APM. 

This implements all of the outlined acceptance criteria, there's also one extra addition in that the locations we had a "Change configuration" button that used to open the flyout (no data states for example) now navigate to the settings page. Those changes are [all in this commit](https://github.com/elastic/kibana/commit/f4c7b1189717c00574bc65d722395de549295ddc).

<img width="1680" alt="Screenshot 2019-07-30 12 20 08" src="https://user-images.githubusercontent.com/471693/62125208-69ceb780-b2c4-11e9-8dd0-7956b69f5b60.png">

One thing to note is that I've had to disable some functional tests for now. I am still currently looking in to those, but as that was the only thing blocking this PR being opened and reviewed I thought it best to disable them. **

===============================================

** The issues I'm seeing in the functional tests are very odd. For example, in one of the tests this assertion fails:

```
const logStreamEntries = await infraLogStream.getStreamEntries();
expect(logStreamEntries.length).to.be.greaterThan(0);
```

As `logStreamEntries.length` returns with a length of `0`. However, when I check the screenshot output in the failures folder the elements are rendered in full. Similarly, if I check the HTML output in the `failure_debug` folder the DOM elements with the correct `data-test-subj` all exist. So I can't possibly see how they're `0`.

Even more curious if I add a `console.log` (and nothing else), e.g.:

```
const logStreamEntries = await infraLogStream.getStreamEntries();
console.log( logStreamEntries);
expect(logStreamEntries.length).to.be.greaterThan(0);
```

then there's no longer a length of `0`, the items exist, and the test passes.

It's very weird, and very flaky 🙈

===============================================

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

elasticmachine commented:
Pinging @elastic/infra-logs-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9955/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9981/)

Kerry350 commented:
I'm going to have to throw in the towel on the functional tests that are failing (it's a subset of them, not all). I've commented them out for now so the build can pass. I'm 100% at a loss in debugging the issue, and I've actually sunk more time into looking at that than it took to implement all the functionality 🙈 Two of them I didn't look at in detail, as I was looking so in detail at the one outlined in the description. 

Fails:
```
const logStreamEntries = await infraLogStream.getStreamEntries();
expect(logStreamEntries.length).to.be.greaterThan(0);
```

Passes:
```
const logStreamEntries = await infraLogStream.getStreamEntries();
console.log(logStreamEntries);
expect(logStreamEntries.length).to.be.greaterThan(0);
```

It suggests *something* is happening during that small amount of execution time in the `console.log`, but then a "fix" based purely on that knowledge would just be susceptible to race conditions and flakiness anyway 🤷‍♀️

If I'm missing something / someone has seen something like this then I'm definitely all ears!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10204/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10216/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10239/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10399/)

Zacqary commented on x-pack/legacy/plugins/infra/public/components/source_configuration/view_source_configuration_button.tsx:
```diff
@@ -0,0 +1,40 @@
+      key={href}
+      path={href}
+      children={({ match, history }) => (
+        <EuiButton data-test-subj={dataTestSubj} color="primary" onClick={() => history.push(href)}>
```
Maybe `useMemo` this `onClick` with the value of `href`?

Kerry350 replied:
To use `useMemo` or `useCallback` here would mean moving that definition to the top of the component function scope, and there it's difficult to access `history`. I don't think in this instance there's much (if anything) to be gained by memoizing the handler - referential equality aside inline functions are rarely a bottleneck, and this component won't be re-rendering a lot (if at all).

Zacqary commented on x-pack/legacy/plugins/infra/public/pages/logs/stream/page_header.tsx:
```diff
@@ -0,0 +1,52 @@
+}
+
+export const StreamPageHeader = injectUICapabilities(
+  injectI18n((props: StreamPageHeaderProps) => {
```
We're not supposed to be using `injectI18N` anymore: https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/GUIDELINE.md#in-reactjs

Existing files with `inject` are no big deal IMO, but ion new code let's replace with `i18n.translate` as recommended.

Kerry350 replied:
Will fix 👍

Zacqary commented on x-pack/legacy/plugins/infra/public/pages/shared/settings/index.tsx:
```diff
@@ -0,0 +1,22 @@
+}
+
+export const SettingsPage = injectUICapabilities(
+  injectI18n(({ intl, uiCapabilities }: SettingsPageProps) => (
```
Same as above, use `i18n.translate`. Although I don't actually see where `intl` is being used in this component?

Kerry350 replied:
Will fix 👍

Kerry350 commented:
@Zacqary Thanks for the review so far 👍 

Regarding these two points:

> `:focus` styles are weird on the Stream tab, it cuts off the left border.

> Is there any way we can reduce the delay on switching tabs?

These are actually both issues that stem from me copying the SIEM implementation exactly (if you navigate to the SIEM app you'll see the styles and the delay). 

The focus style comes from the `<EuiLink />`. And the delay stems from the fact the hierarchy in that implementation used an `<EuiLink />` wrapping the `<EuiTab>`, using a "proper" link in the form of `<EuiLink />` causes a delay as we hit the top level Angular router first.

In the original `<RoutedTabs />` implementation I'd used `<EuiTab />` only and called `history.push` in the `onClick` handler - which was a lot more performant as we just hit our `react-router` instance.

I agree these are both unpleasant, and essentially a regression.

I'd say there's three ways to proceed here:

1) Ship with exactly the same implementation (which has these issues)
2) Ship with a visually similar but functionally improved implementation (whereby I would look to bypass using `<EuiLink />` again, but make the tabs look like the links).
3) Attempt to share exactly the same implementation between apps via the shared component functionality.

Whilst 3 sounds the best, I have big reservations about it because, whilst we're striving for the same visuals, the underlying functionality here falls down to routing and navigation and that's a bit of a nightmare to share precisely between three apps (Infra, SIEM, and APM). I think we're all using `react-router`, but even then things differ in how we handle navigation. Furthermore, we have a ticket coming up where certain tabs will be blocked and not rendered based on permissions - and for that I'd like to extend the functionality of `<RoutedTabs />`, and I doubt it will translate 1:1 with the other apps.

So I'll proceed with 2 which removes these problems, but allows us to share the look and feel of the navigation.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10538/)

Kerry350 commented:
@Zacqary I've addressed your feedback so far:

The navigation is now more performant when switching, and the visual discrepencies are cleaned up.

![navigation](https://user-images.githubusercontent.com/471693/62364116-a692ec80-b518-11e9-91d4-4e96868420fa.gif)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10602/)

Zacqary commented:
@Kerry350 Looks great! Are you able to take it a step further and prevent the logs stream from having to reload data when switching away from the Settings tab?

If it's very complicated to fix, that might not be the biggest issue. Users will probably be changing settings when they switch to the Settings tab which would require a stream reload anyway. Unless persisting the logs data when you switch to the Settings tab would also allow it to start loading new data before you even switch away from the tab?

Zacqary commented:
Can we also get some kind of "Are you sure?" prompt if you make changes, then click the Stream tab instead of clicking Apply? Otherwise those changes are lost

Kerry350 commented:
@Zacqary 

> Are you able to take it a step further and prevent the logs stream from having to reload data when switching away from the Settings tab?

If it's very complicated to fix, that might not be the biggest issue. Users will probably be changing settings when they switch to the Settings tab which would require a stream reload anyway. Unless persisting the logs data when you switch to the Settings tab would also allow it to start loading new data before you even switch away from the tab?

This would be a very involved and non-trivial change, and I doubt we want to focus time in that direction at the moment given the feature priorities. 

I realise that comes with an air of negativity so I wanted to highlight architecturally why it's not a simple improvement:

- Historically we don't use caching functionality. Log entries are served - data wise - by a combination of GraphQL and Redux. On the client GraphQL side we turn off caching with a `fetchPolicy` of `no-cache`, the reason for this is Apollo's caching functionality was found to have quirks / bugs. We don't cache within Redux for later consumption either. Adding on to that, we want to, very soon, remove GraphQL and Redux usage from the locations they still exist. So we don't, as it stands, have an easy way to instantly re-access data that was previously queried for display.

- Every independent page (currently) handles it's data in an independent way. When the `stream` page mounts it fetches and displays it's data, the same happens for `settings` (+ `inventory` and `metrics-explorer`). They don't share anything in any way anymore, this was different with the flyout as it was rendered *within* every page. This loops back in to either having a form of caching mechanism, or hoisting some sort of data sharing mechanism to the `index` page (also not in place).

- We don't actually persist last used query parameters when navigating between tabs, so when you go to `/settings` and then back to `/stream`, when the `stream` page re-mounts it will set itself back up with a new point in time etc.

If this was something we wanted to look at long term - now that we have more "pages" coming in and more navigation needs - I'd say we need to look at this as a whole team and agree upon an approach for the whole of infrastructure + logs. Which, I personally feel, wouldn't make sense until we've got everything moved over to the "simple HTTP API" approach. Even in that scenario, for caching, we can't make use of something simple like regular HTTP cache headers as we use `POST`.

Kerry350 commented:
@Zacqary 

> Can we also get some kind of "Are you sure?" prompt if you make changes, then click the Stream tab instead of clicking Apply? Otherwise those changes are lost

This would be a nice addition, I'll look at this now 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10810/)

Kerry350 commented:
@Zacqary I've added in a prompt for when changes are about to be lost. I've used the `<Prompt />` component that ships directly with `react-router`. It works well, the only downside (well, that's up for debate) is it uses a native browser prompt, so it's not stylised at all. I don't think this is an issue, given how rarely it will probably be seen.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10857/)

Kerry350 commented:
Awesome, thanks for the reviews. 

Just a note to say I've got a small followup PR that will come with a fix for the breadcrumbs being incorrect sometimes. It wasn't broken by this work, it exists in Master, but I spotted it as part of the navigation work. But I don't want to block this big ol' chunk getting merged.
